libmachine/provision/debian.go

```diff
@@ -0,0 +1,213 @@
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"text/template"
+
+	"github.com/docker/machine/drivers"
+	"github.com/docker/machine/libmachine/auth"
+	"github.com/docker/machine/libmachine/engine"
+	"github.com/docker/machine/libmachine/provision/pkgaction"
+	"github.com/docker/machine/libmachine/swarm"
+	"github.com/docker/machine/log"
+	"github.com/docker/machine/utils"
+)
+
+func init() {
+	Register("Debian", &RegisteredProvisioner{
+		New: NewDebianProvisioner,
+	})
+}
+
+func NewDebianProvisioner(d drivers.Driver) Provisioner {
+	return &DebianProvisioner{
+		GenericProvisioner{
+			DockerOptionsDir:  "/etc/docker",
+			DaemonOptionsFile: "/etc/systemd/system/docker.service",
+			OsReleaseId:       "debian",
+			Packages: []string{
+				"curl",
+			},
+			Driver: d,
+		},
+	}
+}
+
+type DebianProvisioner struct {
+	GenericProvisioner
+}
+
+func (provisioner *DebianProvisioner) Service(name string, action pkgaction.ServiceAction) error {
+	// daemon-reload to catch config updates; systemd -- ugh
+	if _, err := provisioner.SSHCommand("sudo systemctl daemon-reload"); err != nil {
+		return err
+	}
+
+	command := fmt.Sprintf("sudo systemctl %s %s", action.String(), name)
+
+	if _, err := provisioner.SSHCommand(command); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (provisioner *DebianProvisioner) Package(name string, action pkgaction.PackageAction) error {
+	var packageAction string
+
+	updateMetadata := true
+
+	switch action {
+	case pkgaction.Install:
+		packageAction = "install"
+	case pkgaction.Remove:
+		packageAction = "remove"
+		updateMetadata = false
+	case pkgaction.Upgrade:
+		packageAction = "upgrade"
+	}
+
+	switch name {
+	case "docker":
+		name = "lxc-docker"
+	}
+
+	if updateMetadata {
+		if _, err := provisioner.SSHCommand("sudo apt-get update"); err != nil {
+			return err
+		}
+	}
+
+	command := fmt.Sprintf("DEBIAN_FRONTEND=noninteractive sudo -E apt-get %s -y  %s", packageAction, name)
+
+	log.Debugf("package: action=%s name=%s", action.String(), name)
+
+	if _, err := provisioner.SSHCommand(command); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (provisioner *DebianProvisioner) dockerDaemonResponding() bool {
+	if _, err := provisioner.SSHCommand("sudo docker version"); err != nil {
+		log.Warnf("Error getting SSH command to check if the daemon is up: %s", err)
+		return false
+	}
+
+	// The daemon is up if the command worked.  Carry on.
+	return true
+}
+
+func (provisioner *DebianProvisioner) Provision(swarmOptions swarm.SwarmOptions, authOptions auth.AuthOptions, engineOptions engine.EngineOptions) error {
+	provisioner.SwarmOptions = swarmOptions
+	provisioner.AuthOptions = authOptions
+	provisioner.EngineOptions = engineOptions
+
+	if provisioner.EngineOptions.StorageDriver == "" {
+		provisioner.EngineOptions.StorageDriver = "aufs"
+	}
+
+	// HACK: since debian does not come with sudo by default,
+	// we check if present and install first
+	out, err := provisioner.SSHCommand("which sudo")
```

Comment:
Any reason not to just run the `apt-get update && apt-get install -y sudo` directly?  It will be a little extra time, but that's mostly dwarfed by the total time cost of creating and configuring a machine, and it would reduce the code complexity.